pkgs/development/python-modules/flask-admin/default.nix

```diff
@@ -22,9 +23,16 @@
 , sqlalchemy-utils
 , wtf-peewee
 , wtforms
-}:
-
-buildPythonPackage rec {
+}: let
+  wtforms2 = wtforms.overridePythonAttrs (old: rec {
+    version = "2.3.3";
+    src = old.src.override {
+      inherit version;
+      sha256 = "81195de0ac94fbc8368abbaf9197b88c4f3ffd6c2719b5bf5fc9da744f3d829c";
+    };
+    doCheck = false;
+  });
```

Comment:
We shouldn't do this. It will cause issues which will be hard to debug.

Reply:
so... leave it broken or what do you want here?
buku appears to be the only user of this module, so the override could stay isolated to that.

Reply:
yes for python modules (as opposed to leaf packages) it's not possible to override their dependencies. So by bumping wtforms to 3, flask-admin was broken without solution. The only solution to prevent that is not to merge breaking updates in the first place. Please consider that next time before merging.

Reply:
Would it help, if we add the version bump of WTForms as an extra package `wtforms3`? So everything that still relies on WTForms2 could use the original package, any any package needing the version 3, could use `wtforms3`?

Reply:
> Would it help, if we add the version bump of WTForms as an extra package `wtforms3`? So everything that still relies on WTForms2 could use the original package, any any package needing the version 3, could use `wtforms3`?

If they both end up in the same pythonEnv things still break because they share the same module name internally.

Reply:
so what is the recommended way of updating python packages, if new versions are required? I can imagine, that some projects don't bump their dependencies and break, when new versions of their dependencies get introduced.. Like it happening here? 
Do we wait (indefinitely perhaps) for the project owners to update their code? I can understand @symphorien reasoning here. But to include [pgadmin4](https://github.com/NixOS/nixpkgs/pull/154764) to nixpkgs, the package needs the new version of `wtforms`. I really wouldn't like to wait for `flask-admin`s owner to update their code, so I can use `wtforms` to finally include `pgadmin4` in nixpkgs.

Reply:
> yes for python modules (as opposed to leaf packages) it's not possible to override their dependencies

That's really not true? It causes build failures if conflicting versions are pulled into a single closure from multiple indirect dependencies, but that's a very solvable problem and a far cry from "not possible". Though to be fair, it does become a little painful if the dependency isn't particularly transient and is commonly found in other dependencies (or even dependent packages themselves).

Yes, the python story in nixpkgs sucks for a number of reasons, but I don't really think that's a good reason to avoid fixing things where we can..? A working package with minor caveats to be aware of still beats an entirely broken one.

> It will cause issues which will be hard to debug.

For anyone dealing with writing python derivations, I'd expect them to already be used to seeing the "multiple versions of X found" error tbh, so I'd argue that "hard to debug" isn't really true. It's a pretty self-explanatory error message, and so you reach for `nix why-depends` to pinpoint exactly what the problem is, no?

> But to include pgadmin4 to nixpkgs, the package needs the new version of wtforms

It would honestly have made more sense to just override `wtforms` locally in the pgadmin4 derivation. AFAICT it's primarily a program (which makes it a "leaf package" @symphorien is referring to above), so doing so is unlikely to cause issues or conflicts in most closures.

Reply:
> so what is the recommended way of updating python packages, if new versions are required?

There is none. We search for patches and break and mark broken what is considered less useful or try to move none library packages out of pythonPackages because there we can do whatever we want.

> A working package with minor caveats to be aware of still beats an entirely broken one.

The problem is that the breakages are silent and you notice three weeks later than one package is empty and still behaves like it was installed because a dependency draw in the same package in another version.

> It would honestly have made more sense to just override wtforms locally in the pgadmin4 derivation. AFAICT it's primarily a program (which makes it a "leaf package" @symphorien is referring to above), so doing so is unlikely to cause issues or conflicts in most closures.

Yeah, we can do that. That should be no problem.

Reply:
> That's really not true? It causes build failures if conflicting versions are pulled into a single closure from multiple indirect dependencies, but that's a very solvable problem and a far cry from "not possible". 

Indeed it's technically possible, but it is forbidden in nixpkgs by policy afaik. To make things manageable.

Reply:
> The problem is that the breakages are silent and you notice three weeks later than one package is empty and still behaves like it was installed because a dependency draw in the same package in another version.

I don't understand what you're referring to, can you expand on this scenario? I'm not sure what an "empty" package means. The nixpkgs build is expected to fail if version mismatches occur due to future changes, and of course once `flask-admin` is properly updated then this override would be obviously unneeded and removed. I'm not really aware of python version overrides being uncommon practice in nixpkgs when needed, but the impact of such an override does vary depending on how the dependency is used to be fair.

Reply:
Maybe we don't have to have an override here. As mentioned [here](https://github.com/flask-admin/flask-admin/issues/2215#issuecomment-1021519097), a new release is on its way and will probably be published this weekend. I just tested the current master branch of `flask-admin` and it works like a charm with `wtforms` Version 3.

Reply:
> I'm not sure what an "empty" package means.

A package that contains no files. That happens when pip installs a package to ``$out`` which is already in the environment and then nothing happens except that `$out` is created and some directories and that ends up being the package.

> of course once flask-admin is properly updated then this override would be obviously unneeded and removed.

Lets be real, there is a high chance someone will forget.

> I'm not really aware of python version overrides being uncommon practice in nixpkgs

We nuked most of them from the tree a few months ago so except pytest there should be almost none.

Reply:
> > I'm not sure what an "empty" package means.
> 
> A package that contains no files. That happens when pip installs a package to `$out` which is already in the environment and then nothing happens except that `$out` is created and some directories and that ends up being the package.

Well that's unfortunate as a failure case, but is a much more extraordinary case that isn't relevant here - plus could easily be guarded against if it were a real concern. A package that accidentally depends on itself if it weren't overridden would cause infinite recursion on evaluation, no? A situation where that is narrowly avoided due to a stray override seems like it would be an incredibly exceptional one. You'd also hope installtests would catch it, but .-.

> > of course once flask-admin is properly updated then this override would be obviously unneeded and removed.
> 
> Lets be real, there is a high chance someone will forget.

Sorry, I want to say I highly doubt your statement here, as it would be immediately obvious to anyone looking at the file to update it. Perhaps not to newcomers, though? and then if github doesn't show it in the diff... though, you could apply that as an argument to reject literally any change in nixpkgs, so it doesn't feel very useful?

> > I'm not really aware of python version overrides being uncommon practice in nixpkgs
> 
> We nuked most of them from the tree a few months ago so except pytest there should be almost none.

Gotcha, well it sucks that this was apparently done without any alternatives proposed. Is any of this policy documented somewhere?
(but really, the fact that nixpkgs allows merges to break any package is honestly the real problem here, unfortunately)

Reply:
> is a much more extraordinary case that isn't relevant here

I already got this once into master and unstable for 3 weeks and no one noticed that the checkInput poluted the environment and the package ended up being empty. 🙃

> A package that accidentally depends on itself if it weren't overridden would cause infinite recursion on evaluation, no?

For nix they are two different packages but for python not and we currently do not have any safeguard for this.

> Gotcha, well it sucks that this was apparently done without any alternatives proposed

Limitation of python which cannot easily and universally overcome.

> Is any of this policy documented somewhere?

IDK, probably not.

Reply:
[Flask-admin got updated to version 1.6.0](https://pypi.org/project/Flask-Admin) which includes an update to WTForms 3:

> 1.6.0
> Dropped Python 2 support
> WTForms 3.0 support
> Various fixes

I build it locally and it works flawlessly with WTForms3. The only thing I needed to change was to add `flask_admin/tests/sqla/test_inlineform.py` to disabledTestPaths.

Reply:
Yeah it seems good here, replaced the override with the update!